### PR TITLE
VPN-4850 - Send Glean pings in Android daemon

### PR DESCRIPTION
**Pull request by mcleinman**
## Description

This does not do the keep Gleans used via git submodule in sync with this Android daemon one. 😬 Updating the submodule to 52.7 was giving build errors, and I'm leaving this for another day.

The way I was testing this:
- Build Android, install to device
- Run app, turn on SuperDooper in Developer feature settings
- Turn on app (so that setting reaches the daemon).
- Verify that pings are happening with `adb logcat | grep daemonsession`

## Reference

https://mozilla-hub.atlassian.net/browse/VPN-4850

## Checklist
    
- [X] My code follows the style guidelines for this project
- [X] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [X] I have performed a self review of my own code
- [X] I have commented my code PARTICULARLY in hard to understand areas
- [X] I have added thorough tests where needed

### CODE REVIEW

**Comment by mcleinman** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

**This is just for the `reasons` being added to this ping. The ping itself was previously approved.**

1) What questions will you answer with this data?

When is this ping being sent? For what reason?

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? 

This is foundational to the overall metrics improvement project. 

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

Not having a reason. However, due to some corner cases (like a crashing daemon), we're not able to fully understand the data without these reasons.

4) Can current instrumentation answer these questions?

No, unfortunately.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Name</td>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>N/A</td>
    <td>N/A</td>
    <td>N/A</td>
    <td>N/A</td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

This collection is documented in the Glean Dictionary at https://dictionary.telemetry.mozilla.org/"

7) How long will this data be collected?  Choose one of the following:

* I want to permanently monitor this data. mcleinman@mozilla.com

8) What populations will you measure?

No filters - all channels, countries, and locales, unless the user has opted out of data collection on that device.

9) If this data collection is default on, what is the opt-out mechanism for users?

When launching the app for the first time, a user is given an option for whether to allow data collection. After this initial set up screen, a user can always toggle data collection permissions in the System Preferences screen.

10) Please provide a general description of how you will analyze this data.

Dashboards that product managers and others will consult on a regular basis.

11) Where do you intend to share the results of your analysis?

Within the Mozilla VPN team and Security and Privacy Products team.

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? 

No

**Comment by mcleinman** [comment]
Tagging @travis79 for data review.

**Review by strseb — APPROVED** [review]

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt [hunk 1]
```diff
@@ -32,6 +34,29 @@ class VPNService : android.net.VpnService() {
...
     val mConnectionHealth = ConnectionHealth(this)
     private var mCityname = ""
+    private var mBackgroundPingTimerMSec: Long = 3 * 60 * 60 * 1000 // 3 hours, expressed in milliseconds
+    private val mMetricsTimer: CountDownTimer = object : CountDownTimer(
+        mBackgroundPingTimerMSec,
+        mBackgroundPingTimerMSec / 4
+    ) {
+        override fun onTick(millisUntilFinished: Long) {}
+        override fun onFinish() {
+            Log.i(tag, "Sending daemon_timer ping")
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt:44]: Does it include the reason? It didn't seem like that in the logs. Happy to remove this if it include the reason.

File: src/apps/vpn/telemetry/pings.yaml [hunk 2]
```diff
@@ -31,11 +31,21 @@ daemonsession:
...
     - vpn-telemetry@mozilla.com
     - mcleinman@mozilla.com
+  reasons:
+    daemon_flush: >
+      Sent right before a session start metric is recorded. This should have
+      only a UUID that appears in no other pings.
+    daemon_start: >
+      Sent right after a session start metric is recorded.
+    daemon_timer: >
+      (Android only) Sent every three hours during an active VPN session.
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/telemetry/pings.yaml:49]: I'm still not confident that iOS won't sometimes pause process (and thus the timer would be unreliable). From yesterday (and before), I do feel confident we get some processing - especially when there are active network requests happening.

I'm inclined to keep this unless we decide to change the current tech spec.

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt [hunk 3]
```diff
@@ -32,6 +34,29 @@ class VPNService : android.net.VpnService() {
...
+            // send ping
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt]: removed

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt [hunk 4]
```diff
@@ -32,6 +34,29 @@ class VPNService : android.net.VpnService() {
...
+            if (isSuperDooperMetricsActive) {
+                Pings.daemonsession.submit(
+                    Pings.daemonsessionReasonCodes.daemonTimer
+                )
+            }
+            // reset timer
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt]: removed

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt [hunk 5]
```diff
@@ -32,6 +34,29 @@ class VPNService : android.net.VpnService() {
     private var mAlreadyInitialised = false
     val mConnectionHealth = ConnectionHealth(this)
     private var mCityname = ""
+    private var mBackgroundPingTimerMSec: Long = 3 * 60 * 60 * 1000 // 3 hours, expressed in milliseconds
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt]: updated

**Comment by mcleinman** [comment]
> > This does not do the keep Gleans used via git submodule in sync with this Android daemon one.
> 
> FWIW, keeping this in sync this will not be a problem once #6960 is merged. With merging that the submodule update will be required here though. [UPDATE] That is merged, please rebase this PR otherwise it will fail on main after the merge.
> 
> Also, would you mind updating the title of the pull request?

Merged in main, and updated the title of this request. Thanks!

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (brizental) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt:44]: No reason. I think this is redundant regardless, but no strong opinion.

File: src/apps/vpn/telemetry/pings.yaml [hunk 6]
```diff
@@ -31,11 +31,21 @@ daemonsession:
     iOS Network Extension. This ping will record start/end datestamp,
     session-based metrics, etc.
   include_client_id: true
-  send_if_empty: false
+  send_if_empty: true
```
[the remark this answers is not in the document]
Comment (brizental) [reply src/apps/vpn/telemetry/pings.yaml:34]: I see. This is useful for documentation purposes though. We don't expect this ping to be empty, so this should be false.

File: src/apps/vpn/telemetry/pings.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (brizental) [reply src/apps/vpn/telemetry/pings.yaml:49]: > I'm still not confident that iOS won't sometimes pause process 

We can't be confident about this on Android either. No strong push. Let's leave it as is and address this separately in the future.

File: android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcleinman) [reply android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNService.kt:44]: I can see a world where we're trying to figure out some weird bug, and the reason is helpful. Going to leave this in and resolve.

File: src/apps/vpn/telemetry/pings.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/telemetry/pings.yaml:34]: Fair, updated.

**Pull request merged**